librad/src/git/storage.rs

```diff
@@ -755,6 +755,32 @@ where
             |peer| self.certifiers_of(&urn, &peer),
         )?;
 
+        // Ensure refs/namespaces/<certifier>/remotes/<tracked>/rad/id exists
+        // and adopt it as our own.
+        //
+        // FIXME(haxpenny): Currently rad/ids/* doesn't exist if something is signed by
+        // OwnedKey. Really looking forward to identity documents at this rate.
+        {
+            for certifier in self.certifiers_of(&urn, &remote_peer)?.into_iter() {
+                let remote_certifier =
+                    NamespacedRef::rad_id(certifier.id.clone()).with_remote(remote_peer);
+
+                if !self.has_ref(&remote_certifier)? {
+                    return Err(Error::MissingCertifier { certifier, urn });
+                }
+
+                let local_certifier = remote_certifier.clone().with_remote(None);
+
+                // We may have already cloned this certifier at some point
+                if !self.has_ref(&local_certifier)? {
+                    remote_certifier
+                        .symbolic_ref(local_certifier, Force::False)
```

Comment:
If I didn't get `SymbolicRef` backwards, this will create:

    r/nss/<certifier>/refs/rad/id -> r/nss/<project>/refs/remotes/<remote peer>/rad/ids/<certifier>

Below we're creating the symref in the opposite direction (but from our own namespace):

    r/nss/<project>/refs/rad/ids/<certifier> -> r/nss/<certifier>/refs/rad/id

Because we want `<project>` to always point to the latest of `<certifier>`.

Now, creating the first symref is not technically _wrong_, because fetching `<certifier>` will replace it with a direct ref when there are updates. It may be a bit confusing, though.

Reply:
Well, I'm glad you brought this up because that code is also `parse`'ing the wrong thing :sweat_smile: It's trying to get a `Hash` from `refs/namespaces/hwd1yre8itytmdeqt8ezb1f5sbtcz1p8z3cqy66nbyh73yh9idqc43k3ujy/refs/remotes/hydneauugdxtsrpg5o636gizgmoxoo4ymbkk74am5qf1uxkastoydo/rad/ids/hwd1yregdu7n8rqpsmft43ut9k3b1h9j4bxpaaw9whzk8sua89zn4zqgruc`

Reply:
I'm a little bit lost now -- don't we need to ensure that the `rad/id` (that is, not `remotes/xx/rad/id`) of the top-level namespace exists before symlinking to it?

Reply:
So I believe it goes like this (where `<namespace>` is what we're cloning/fetching and `<remote_peer>` is the peer we're cloning/fetching from):
* [For each tracked `<peer>` and for each `<certifier>` under that tracked peer we fetch](https://github.com/radicle-dev/radicle-link/blob/fintan/top-level-certifier/librad/src/git/types.rs#L298-L306)
  * From: `refs/namespaces/<certifier>/refs[/remotes/<peer>]/rad/id`
  * Into: `refs/namespaces/<certifier>/refs/remotes/<peer>/rad/id`
* [Then we iterate over `refs/namespaces/<namespace>/refs/remotes/<remote_peer>/rad/ids/*` which gives us these identities](https://github.com/radicle-dev/radicle-link/blob/fintan/top-level-certifier/librad/src/git/storage.rs#L762-L770)
* [For each `<certifier>` we create a symref](https://github.com/radicle-dev/radicle-link/blob/fintan/top-level-certifier/librad/src/git/storage.rs#L779-L786):
  * New symbolic ref: `refs/namespaces/<certifier>/rad/id`
  * Points to: `refs/namespaces/<namespace>/refs/remotes/<remote_peer>/rad/ids/<certifier>`

Writing this out, I get the feeling I'm mixing this last step up. Should it be that we create the `refs/namespaces/<certifier>/rad/id` from the `refs/namespaces/<certifier>/refs/remotes/<remote_peer>/rad/id`? I guess what's confusing me is that [original block of code that wasn't firing](https://github.com/radicle-dev/radicle-link/blob/fintan/top-level-certifier/librad/src/git/storage.rs#L758-L790) because of the parsing issue.

Reply:
This seems arse-ways now that I think about. The `rad/ids/*` are always just symref'd, correct? 

So, if I'm getting this right, the original [block here](https://github.com/radicle-dev/radicle-link/blob/master/librad/src/git/storage.rs#L762-L795) was creating a symref to the `refs/namespaces/<certifier>/rad/id`. The issue I'm having here is that we don't actually have that reference, instead we have `refs/namespaces/<certifier>/refs/remotes/<peer>/rad/id`.

Sorry that this is taking so long to sink in for me, it's pretty frustrating :disappointed:

Reply:
Ok, so in the last step, if we **already have** `refs/namespaces/<certifier>/rad/id`, what we want is symrefs

    refs/namespaces/<current project>/rad/ids/<certifier> -> refs/namespaces/<certifier>/rad/id

This ensures that forall `<project>`s, we get the updated tips by fetching `refs/namespaces/<certifier>` automatically. Note that we're not touching any `refs/remotes/`!

If we **do not have** `refs/namespaces/<certifier>/rad/id`, we **should** run the verification for `refs/namespaces/<certifier>/refs/remotes/<remote peer>/rad/id`, and **iff** that checks out, create a **direct** ref `refs/namespaces/<certifier>/rad/id`. Because there is no proper verification, we just pretend it's valid and create that direct ref.

Reply:
7th times a charm :pray: